File: streamlit_app/services/board_service.py
"""팀 게시판 서비스 - 로컬 JSON 저장소."""
from __future__ import annotations

from services import local_store
import time
import hashlib


def _hash_password(password: str) -> str:
    """비밀번호 해시."""
    return hashlib.sha256(password.encode()).hexdigest()


def can_read_secret(post: dict, user_id: str, team: dict | None) -> bool:
    """비밀글 읽기 권한 확인. 작성자 본인 또는 운영진이면 True."""
    if not post.get("isSecret"):
        return True
    if post.get("authorId") == user_id:
        return True
    if team:
        if team.get("leaderId") == user_id:
            return True
        if user_id in team.get("admins", []):
            return True
    return False


def can_manage_post(post: dict, user_id: str, team: dict | None) -> bool:
    """게시글 수정/삭제 권한 확인. 작성자 본인 또는 운영진."""
    if post.get("authorId") == user_id:
        return True
    if team:
        if team.get("leaderId") == user_id:
            return True
        if user_id in team.get("admins", []):
            return True
    return False


def create_post(
    title: str,
    content: str,
    author_id: str,
    author_name: str,
    team_id: str,
    is_secret: bool = False,
    password: str = "",
) -> dict | None:
    """게시글 작성."""
    post_data = {
        "title": title,
        "content": content,
        "authorId": author_id,
        "authorName": author_name,
        "teamId": team_id,
        "isSecret": is_secret,
        "passwordHash": _hash_password(password) if is_secret and password else "",
        "createdAt": int(time.time() * 1000),
        "updatedAt": int(time.time() * 1000),
    }
    post_id = local_store.add_doc("board_posts", post_data)
    post_data["id"] = post_id
    return post_data


def get_team_posts(team_id: str) -> list:
    """팀 게시글 목록 조회 (최신순)."""
    return local_store.query(
        "board_posts",
        [("teamId", "==", team_id)],
        order_by="createdAt",
        order_dir="DESC",
        limit=50,
    )


def get_post(post_id: str) -> dict | None:
    """단일 게시글 조회."""
    doc = local_store.get_doc("board_posts", post_id)
    if doc:
        doc["id"] = post_id
        return doc
    return None


def verify_post_password(post_id: str, password: str) -> bool:
    """비밀글 비밀번호 확인."""
    post = get_post(post_id)
    if not post:
        return False
    if not post.get("isSecret"):
        return True
    return post.get("passwordHash", "") == _hash_password(password)


def update_post(
    post_id: str,
    title: str,
    content: str,
    is_secret: bool = False,
    password: str = "",
) -> bool:
    """게시글 수정."""
    update_data = {
        "title": title,
        "content": content,
        "isSecret": is_secret,
        "updatedAt": int(time.time() * 1000),
    }
    if is_secret and password:
        update_data["passwordHash"] = _hash_password(password)
    elif not is_secret:
        update_data["passwordHash"] = ""
    local_store.update_doc("board_posts", post_id, update_data)
    return True


def delete_post(post_id: str) -> None:
    """게시글 삭제."""
    local_store.delete_doc("board_posts", post_id)

File: src/services/boardService.ts
import {
  doc, collection, addDoc, getDocs, getDoc, deleteDoc, updateDoc,
  query, where, orderBy, limit
} from 'firebase/firestore';
import { getFirebaseFirestore } from '../utils/firebase';
import { getCached, invalidateCache } from '../utils/firestoreCache';
import type { BoardPost } from '../types';

function hashPassword(password: string): string {
  // Simple hash for client-side — matches Python's SHA256
  // In production, use a proper crypto library
  let hash = 0;
  for (let i = 0; i < password.length; i++) {
    const char = password.charCodeAt(i);
    hash = ((hash << 5) - hash) + char;
    hash = hash & hash;
  }
  return Math.abs(hash).toString(16);
}

export async function createPost(
  title: string,
  content: string,
  authorId: string,
  authorName: string,
  teamId: string,
  isSecret: boolean = false,
  password: string = ''
): Promise<BoardPost> {
  const fs = getFirebaseFirestore()!;
  const postData: Omit<BoardPost, 'id'> = {
    title,
    content,
    authorId,
    authorName,
    teamId,
    isSecret,
    passwordHash: isSecret && password ? hashPassword(password) : '',
    createdAt: Date.now(),
    updatedAt: Date.now(),
  };
  const ref = await addDoc(collection(fs, 'board_posts'), postData);
  invalidateCache(`posts:${teamId}`);
  return { id: ref.id, ...postData };
}

export async function getTeamPosts(teamId: string): Promise<BoardPost[]> {
  return getCached(`posts:${teamId}`, async () => {
    const fs = getFirebaseFirestore()!;
    const q = query(
      collection(fs, 'board_posts'),
      where('teamId', '==', teamId),
      orderBy('createdAt', 'desc'),
      limit(50)
    );
    const snap = await getDocs(q);
    return snap.docs.map((d) => ({ id: d.id, ...d.data() }) as BoardPost);
  }, 3 * 60 * 1000);
}

export async function getPost(postId: string): Promise<BoardPost | null> {
  const fs = getFirebaseFirestore()!;
  const snap = await getDoc(doc(fs, 'board_posts', postId));
  if (!snap.exists()) return null;
  return { id: snap.id, ...snap.data() } as BoardPost;
}

export async function updatePost(
  postId: string,
  title: string,
  content: string,
  isSecret: boolean = false,
  password: string = ''
): Promise<void> {
  const fs = getFirebaseFirestore()!;
  const updateData: Record<string, unknown> = {
    title,
    content,
    isSecret,
    updatedAt: Date.now(),
  };
  if (isSecret && password) {
    updateData.passwordHash = hashPassword(password);
  } else if (!isSecret) {
    updateData.passwordHash = '';
  }
  await updateDoc(doc(fs, 'board_posts', postId), updateData);
  invalidateCache('posts:');
}

export async function deletePost(postId: string): Promise<void> {
  const fs = getFirebaseFirestore()!;
  await deleteDoc(doc(fs, 'board_posts', postId));
  invalidateCache('posts:');
}

export function canReadSecret(
  post: BoardPost,
  userId: string,
  team: { leaderId: string; admins?: string[] } | null
): boolean {
  if (!post.isSecret) return true;
  if (post.authorId === userId) return true;
  if (team) {
    if (team.leaderId === userId) return true;
    if (team.admins?.includes(userId)) return true;
  }
  return false;
}

export function canManagePost(
  post: BoardPost,
  userId: string,
  team: { leaderId: string; admins?: string[] } | null
): boolean {
  if (post.authorId === userId) return true;
  if (team) {
    if (team.leaderId === userId) return true;
    if (team.admins?.includes(userId)) return true;
  }
  return false;
}
